src/ar_gateway_middleware.erl

```diff
@@ -237,11 +253,122 @@ get_tx_from_hash_1(TXID) ->
 		Filename -> {ok, TXID, Filename}
 	end.
 
-serve_tx(Filename, Req) ->
+serve_tx(Filename, Path, Fallback, Req, Env) ->
 	TX = ar_storage:read_tx_file(Filename),
-	ContentType = proplists:get_value(<<"Content-Type">>, TX#tx.tags, "text/html"),
-	Headers = #{<<"content-type">> => ContentType},
+	ContentType =
+		case lists:keyfind(<<"Content-Type">>, 1, TX#tx.tags) of
+			{<<"Content-Type">>, V} -> V;
+			false -> <<"text/html">>
+		end,
+	case {ContentType, Path} of
+		{?MANIFEST_CONTENT_TYPE, <<>>} ->
+			serve_manifest(TX, Req);
+		{?MANIFEST_CONTENT_TYPE, _} ->
+			serve_manifest_path(TX, Path, Fallback, Req, Env);
+		{_, <<>>} ->
+			serve_plain_tx(TX, ContentType, Req);
+		_ ->
+			bad_request(Req)
+	end.
+
+serve_manifest(TX, Req) ->
+	case ar_serialize:json_decode(TX#tx.data, [return_maps]) of
+		{ok, #{ <<"index">> := #{ <<"path">> := Index } }} ->
+			serve_manifest_index(Index, Req);
+		{ok, #{ <<"paths">> := Paths }} ->
+			serve_manifest_listing(ar_util:encode(TX#tx.id), maps:keys(Paths), Req);
+		{error, _} ->
+			misdirected_request(Req)
+	end.
+
+serve_manifest_index(Index, Req) ->
+	Hostname = cowboy_req:host(Req),
+	Path = cowboy_req:path(Req),
+	NewPath = filename:join(Path, Index),
+	Location = <<
```

Comment:
Is there a way to serve index without changing the path?

Reply:
We should be able to add this in the future without breaking backwards compatibility, so it should be fine to move forward with this current implementation 🙂

Reply:
👍